### ISSUE TITLE
Inefficient instructions generated for avx512 integer comparison.

### ISSUE THREAD

**Issue by bpickering** [ispc/ispc#1608]
While compiling for an `avx512knl-i32x16` target it appears sub-optimal code is being generated for simple comparisons of varying `int` types (`float` comparisons appear to use efficient instructions).  The issue is occurring with the 1.12.0 binary but can also be reproduced on Godbolt with multiple ispc versions.  Examples below illustrate `float` and `int` cases.

`float` code:
```
export void float_compare(uniform float vin[], uniform float cmp[], uniform float vout[],
                   uniform int count) {
    foreach (index = 0 ... count) {
        varying float v = vin[index];
        varying float c = cmp[index];

        if (c < 0.0)
            v = v * v;
        else
            v = sqrt(v);

        vout[index] = v;
    }
}
```
`float` assembly (foreach full body section):
```
.LBB1_3:                                # %foreach_full_body
        cltq
        vmovups (%rdi,%rax), %zmm1
        vcmpnleps       (%rsi,%rax), %zmm0, %k1
        kortestw        %k1, %k1
        je      .LBB1_5
        vmulps  %zmm1, %zmm1, %zmm1 {%k1}
.LBB1_5:                                # %foreach_full_body
        jb      .LBB1_7
        vsqrtps %zmm1, %zmm2
        vmovaps %zmm1, %zmm2 {%k1}
        vmovaps %zmm2, %zmm1
.LBB1_7:                                # %foreach_full_body
        vmovups %zmm1, (%rdx,%rax)
        addl    $16, %r9d
        addl    $64, %eax
        cmpl    %r8d, %r9d
        jl      .LBB1_3
        cmpl    %ecx, %r9d
        jl      .LBB1_9
```

`int` code:
```
export void int_compare(uniform float vin[], uniform int cmp[], uniform float vout[],
                   uniform int count) {
    foreach (index = 0 ... count) {
        varying float v = vin[index];
        varying int   c = cmp[index];

        if (c < 0)
            v = v * v;
        else
            v = sqrt(v);

        vout[index] = v;
    }
}
```
`int` assembly (foreach full body section):
```
.LBB2_5:                                # %foreach_full_body
        movslq  %r10d, %r10
        vmovups (%rdi,%r10), %zmm2
        vpsrad  $31, (%rsi,%r10), %zmm3
        vpmovdb %zmm3, %xmm3
        vpcmpeqb        %xmm0, %xmm3, %xmm4
        vmovdqa64       %zmm4, %zmm5
        vpternlogq      $15, %zmm4, %zmm4, %zmm5
        vpmovmskb       %xmm5, %eax
        testw   %ax, %ax
        je      .LBB2_6
        vmulps  %zmm2, %zmm2, %zmm5
        vpmovsxbd       %xmm4, %zmm4
        vptestmd        %zmm4, %zmm4, %k1
        vmovaps %zmm2, %zmm5 {%k1}
        vmovaps %zmm5, %zmm2
.LBB2_6:                                # %safe_if_after_true
        vpcmpeqb        %xmm1, %xmm3, %xmm3
        vmovdqa64       %zmm3, %zmm4
        vpternlogq      $15, %zmm3, %zmm3, %zmm4
        vpmovmskb       %xmm4, %eax
        testw   %ax, %ax
        je      .LBB2_8
        vsqrtps %zmm2, %zmm4
        vpmovsxbd       %xmm3, %zmm3
        vptestmd        %zmm3, %zmm3, %k1
        vmovaps %zmm2, %zmm4 {%k1}
        vmovaps %zmm4, %zmm2
.LBB2_8:                                # %if_done
        vmovups %zmm2, (%rdx,%r10)
        addl    $16, %r9d
        addl    $64, %r10d
        cmpl    %r8d, %r9d
        jl      .LBB2_5
        cmpl    %ecx, %r9d
        jge     .LBB2_3
```

The expected output for the `int` case would replace the `vcmpnleps` instruction with a `vpcmpd` variant.   Note the `avx512skx-i32x16` target also seems affected.

**Comment by bpickering** [comment]
Forgot to mention: this seems to occur specifically for the `c < 0` case.  Other `int` comparisons such as `c > 0` or `c == 0` generate `vpcmpltd` and `vpcmpeqd` respectively.

**Comment by dbabokin** [comment]
@DeepakRajendrakumaran Could you please have a look? Looks like LLVM code gen is doing wrong thing.

@bpickering I assume you care about ``avx512skx-i32x16``, not ``avx512knl-i32x16", right?

**Comment by bpickering** [comment]
Actually, current test box is KNL (not using Skylake at present).  I also just found it is possible to work around this issue by incrementing the "c" variable, so the comparison becomes `c < 1` instead of `c < 0`.

**Comment by dbabokin** [comment]
Good to know that KNL target is still in use.

**Comment by DeepakRajendrakumaran** [comment]
@bpickering Sorry about the delayed response. I took a look and my initial impression is it's not a codegen issue. The IR is created with compare for i8 because the original i32 gets truncated to i8 in the IR. I'm not sure why we are truncating the i32. Will update when I have more info.

**Comment by DeepakRajendrakumaran** [comment]
Adding a trimmed test case - https://ispc.godbolt.org/z/7G4vqd

**Comment by DeepakRajendrakumaran** [comment]
The reason for this is inst combine kicking in -

It sees this pattern-
%less_c_load_ = icmp slt <16 x i32> %c_load, zeroinitializer
  %less_c_load__to_boolvec = sext <16 x i1> %less_c_load_ to <16 x i8>

and converts it to a shift right and truncate. Something like this-
 %cmp_load82_offset_load.lobit = ashr <16 x i32> %ptr221_masked_load, <i32 31,......., i32 31>
  %1 = trunc <16 x i32> %cmp_load82_offset_load.lobit to <16 x i8>


This obviously can only be used to check if it's a negative number or not. Hence is not happening for the other comparisons.

**Comment by DeepakRajendrakumaran** [comment]
There are a couple of ways to fix this.

1. LLVM has a fix in trunk which fixes this - https://github.com/llvm/llvm-project/commit/fb5af8b9b93cb270fda7832e24cab9cb62dec16f . I checked this patch on 8.0 and it fixes this.

2. The mask for avx512 being i8 is one reason for this happening due to us inserting truc. We have to consider switching to i32 mask for avx512 and compare performance.

**Comment by dbabokin** [comment]
1. Let's backport the patch to 8.0 and 9.0.
2. The mask in AVX512 is represented as a real mask (in mask registers), but LLVM doesn't have native representation for predicates/masks. So we have to pick the one, which is expected by codegen. Initial design choice to use i8 was made in the early days of AVX512 support and might be the wrong one. We need to check what codegen expects now and adjust.

**Comment by bpickering** [comment]
Is this actually the same issue or closely related to #1609 ?  I posted them separately, but see many of the same ```vpcmpeqb / vptestmd``` instructions being utilized.

**Comment by DeepakRajendrakumaran** [comment]
@bpickering It's not the same issue but it's closely related. The llvm patch will not fix #1609 as it looks for the particular pattern which is relevant for #1608. But the underlying cause for us having the i8 comparisons in the IR in the first place is the i8 mask. I'm looking the IR and assembly and it might be possible for llvm to handle this better too  but ideally looking into the mask size might be needed in the long run

**Comment by DeepakRajendrakumaran** [comment]
I might have jumped the gun on the original bug here. i was working off of a reduced case which was creating the following patterng

ashr-> trunc -> icmp

Looks like the original test creates 
ashr -> trunc -> and -> icmp 

I talked to Craig and he said, he's not sure if that can be handled. His opinion was to to consider i1 masks for avx512 in ISPC.

**Comment by dbabokin** [comment]
i1 sounds reasonable, we should experiment with that.

**Comment by dbabokin** [comment]
Fixed by #1658
